### PR TITLE
Add GPU support to MPI backend

### PR DESCRIPTION
**Pull request by jmerizia**
This adds GPU support to the MPI backend for the following layers:
- [x] all sum reduce
- [x] batchnorm
- [x] transpose
- [ ] ~~conv channel~~
- [x] conv feature
- [ ] ~~conv general~~
- [x] halo exchange
- [ ] ~~interpolate~~*
- [x] linear
- [x] loss
- [x] padnd
- [x] pooling
- [x] sum reduce
- [x] transpose
- [ ] ~~upsampling~~*
- [x] unpadnd

\* will be done in a different PR b/c it involves working with the C++ implementation

### CODE REVIEW

**Comment by rhewett** [comment]
all-sum-reduce, the two convolutions, and linear should be easy to update.  upsample/interpolate may need to be a separate issue.  

I created https://github.com/distdl/distdl/issues/215 for this.  Can you squash down the PR and update the commit message?  I'm looking over the code.

File: src/distdl/nn/padnd.py [hunk 1]
```diff
@@ -1,46 +1,23 @@
...
 
     def __init__(self, pad_width, value):
 
         super(PadNd, self).__init__()
-
         self.pad_width = pad_width
         self.value = value
+        self.torch_pad = self._to_torch_padding(self.pad_width)
+
+    def _to_torch_padding(self, pad):
```
Comment (rhewett) [inline src/distdl/nn/padnd.py:15]: Is this still used?  I thought we removed it.  Should we just remove the layer itself?

Comment (jmerizia) [reply src/distdl/nn/padnd.py:15]: I don't think the layer is being used anywhere, and since it's just a local padding, the only use is for interoperability between how we represent padding, to how PyTorch represents padding.

Somewhat related, I think what would be useful to users down the road is a distributed pad. That's something that can be used in other layers too.

Comment (rhewett) [reply src/distdl/nn/padnd.py:15]: If you take this out, maybe we just remove it in another PR

**Comment by rhewett** [comment]
If you give this a squash, I'll merge it.  We can add the missing ones later.

**Comment by rhewett** [comment]
@jmerizia can you squash this so I can merge it?  Does it have all of the latest bug fixes?
